src/app/clusters/ota-requestor/OTARequestorCluster.cpp

```diff
@@ -0,0 +1,273 @@
+    }
+    return CHIP_NO_ERROR;
+}
+
+void OTARequestorCluster::Shutdown()
+{
+    if (mOtaRequestor)
+    {
+        SuccessOrLog(mOtaRequestor->UnregisterEventHandler(mPath.mEndpointId), SoftwareUpdate,
+                     "Unable to unregister event handling for endpoint %u during shutdown", mPath.mEndpointId);
+    }
+    DefaultServerCluster::Shutdown();
+}
+
+DataModel::ActionReturnStatus OTARequestorCluster::ReadAttribute(const DataModel::ReadAttributeRequest & request,
+                                                                 AttributeValueEncoder & encoder)
+{
+    switch (request.path.mAttributeId)
+    {
+    case OtaSoftwareUpdateRequestor::Attributes::DefaultOTAProviders::Id: {
+        if (!mOtaRequestor)
+        {
+            // There are examples which enable the OTA requestor cluster in their zap configuration but don't
+            // enable the build flag that controls the implementation of the OTA requestor. The behaviour here
+            // maintains backwards compatibility for those examples.
+#if CHIP_DEVICE_CONFIG_ENABLE_OTA_REQUESTOR
+            return CHIP_ERROR_INTERNAL;
+#else
+            return encoder.EncodeEmptyList();
+#endif
+        }
+        return encoder.EncodeList([this](const auto & listEncoder) -> CHIP_ERROR {
+            ProviderLocationList::Iterator providerIterator = mOtaRequestor->GetDefaultOTAProviderListIterator();
+            CHIP_ERROR error                                = CHIP_NO_ERROR;
+            while (error == CHIP_NO_ERROR && providerIterator.Next())
+            {
+                error = listEncoder.Encode(providerIterator.GetValue());
+            }
+            return error;
+        });
+    }
+    case OtaSoftwareUpdateRequestor::Attributes::UpdatePossible::Id:
+        return encoder.Encode(mUpdatePossible);
+    case OtaSoftwareUpdateRequestor::Attributes::UpdateState::Id:
+        if (!mOtaRequestor)
+        {
+            // There are examples which enable the OTA requestor cluster in their zap configuration but don't
+            // enable the build flag that controls the implementation of the OTA requestor. The behaviour here
+            // maintains backwards compatibility for those examples.
+#if CHIP_DEVICE_CONFIG_ENABLE_OTA_REQUESTOR
+            return CHIP_ERROR_INTERNAL;
+#else
+            return encoder.Encode(OtaSoftwareUpdateRequestor::UpdateStateEnum::kUnknown);
+#endif
+        }
+        return encoder.Encode(mOtaRequestor->GetCurrentUpdateState());
+    case OtaSoftwareUpdateRequestor::Attributes::UpdateStateProgress::Id:
+        if (!mOtaRequestor)
+        {
+            // There are examples which enable the OTA requestor cluster in their zap configuration but don't
+            // enable the build flag that controls the implementation of the OTA requestor. The behaviour here
+            // maintains backwards compatibility for those examples.
+#if CHIP_DEVICE_CONFIG_ENABLE_OTA_REQUESTOR
```

Comment:
How about using a different cluster for CHIP_DEVICE_CONFIG_ENABLE_OTA_REQUESTOR and in that case only CodegenIntegration picks what clusters to register?

I.e. we register only either a fully valid cluster or the backwards compatibility cluster, but ifdefs stay in codegen integration.